File: framec_tests/golang/branch/branch.go
// emitted from framec_v0.11.0
// get include files at https://github.com/frame-lang/frame-ancillary-files
package branch
import ( "golang/framelang")

func NewBranch() Branch {
    m := &branchStruct{}
    
    // Validate interfaces
    var _ Branch = m
    var _ Branch_actions = m
    
    // Create and intialize start state compartment.
    m._compartment_ = NewBranchCompartment(BranchState_I)
    
    // Override domain variables.
    m.tape = []string{}
    
    // Send system start event
    e := framelang.FrameEvent{Msg:">"}
    m._mux_(&e)
    return m
}


type BranchState uint

const (
    BranchState_I BranchState = iota
    BranchState_SimpleIf
    BranchState_NegatedIf
    BranchState_Precedence
    BranchState_NestedIf
    BranchState_GuardedTransition
    BranchState_NestedGuardedTransition
    BranchState_F1
    BranchState_F2
    BranchState_F3
)

type Branch interface {
    A() 
    B() 
    C() 
    D() 
    E() 
    F() 
    OnBool(b bool) 
    OnInt(i int16) 
}

type Branch_actions interface {
    log(msg string) 
}


type branchStruct struct {
    _compartment_ *BranchCompartment
    _nextCompartment_ *BranchCompartment
    tape []string
}

//===================== Interface Block ===================//

func (m *branchStruct) A()  {
    e := framelang.FrameEvent{Msg:"A"}
    m._mux_(&e)
}

func (m *branchStruct) B()  {
    e := framelang.FrameEvent{Msg:"B"}
    m._mux_(&e)
}

func (m *branchStruct) C()  {
    e := framelang.FrameEvent{Msg:"C"}
    m._mux_(&e)
}

func (m *branchStruct) D()  {
    e := framelang.FrameEvent{Msg:"D"}
    m._mux_(&e)
}

func (m *branchStruct) E()  {
    e := framelang.FrameEvent{Msg:"E"}
    m._mux_(&e)
}

func (m *branchStruct) F()  {
    e := framelang.FrameEvent{Msg:"F"}
    m._mux_(&e)
}

func (m *branchStruct) OnBool(b bool)  {
    params := make(map[string]interface{})
    params["b"] = b
    e := framelang.FrameEvent{Msg:"OnBool", Params:params}
    m._mux_(&e)
}

func (m *branchStruct) OnInt(i int16)  {
    params := make(map[string]interface{})
    params["i"] = i
    e := framelang.FrameEvent{Msg:"OnInt", Params:params}
    m._mux_(&e)
}

//====================== Multiplexer ====================//

func (m *branchStruct) _mux_(e *framelang.FrameEvent) {
    switch m._compartment_.State {
    case BranchState_I:
        m._BranchState_I_(e)
    case BranchState_SimpleIf:
        m._BranchState_SimpleIf_(e)
    case BranchState_NegatedIf:
        m._BranchState_NegatedIf_(e)
    case BranchState_Precedence:
        m._BranchState_Precedence_(e)
    case BranchState_NestedIf:
        m._BranchState_NestedIf_(e)
    case BranchState_GuardedTransition:
        m._BranchState_GuardedTransition_(e)
    case BranchState_NestedGuardedTransition:
        m._BranchState_NestedGuardedTransition_(e)
    case BranchState_F1:
        m._BranchState_F1_(e)
    case BranchState_F2:
        m._BranchState_F2_(e)
    case BranchState_F3:
        m._BranchState_F3_(e)
    }
    
    if m._nextCompartment_ != nil {
        nextCompartment := m._nextCompartment_
        m._nextCompartment_ = nil
        if nextCompartment._forwardEvent_ != nil && 
           nextCompartment._forwardEvent_.Msg == ">" {
            m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
            m._compartment_ = nextCompartment
            m._mux_(nextCompartment._forwardEvent_)
        } else {
            m._do_transition_(nextCompartment)
            if nextCompartment._forwardEvent_ != nil {
                m._mux_(nextCompartment._forwardEvent_)
            }
        }
        nextCompartment._forwardEvent_ = nil
    }
}

//===================== Machine Block ===================//

func (m *branchStruct) _BranchState_I_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "A":
        compartment := NewBranchCompartment(BranchState_SimpleIf)
        m._transition_(compartment)
        
        return
    case "B":
        compartment := NewBranchCompartment(BranchState_NegatedIf)
        m._transition_(compartment)
        
        return
    case "C":
        compartment := NewBranchCompartment(BranchState_Precedence)
        m._transition_(compartment)
        
        return
    case "D":
        compartment := NewBranchCompartment(BranchState_NestedIf)
        m._transition_(compartment)
        
        return
    case "E":
        compartment := NewBranchCompartment(BranchState_GuardedTransition)
        m._transition_(compartment)
        
        return
    case "F":
        compartment := NewBranchCompartment(BranchState_NestedGuardedTransition)
        m._transition_(compartment)
        
        return
    }
}

func (m *branchStruct) _BranchState_SimpleIf_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "OnBool":
        if e.Params["b"].(bool) {
            m.log("then 1")
        } else {
        }
        if e.Params["b"].(bool) {
        } else {
            m.log("else 1")
        }
        if e.Params["b"].(bool) {
            m.log("then 2")
        } else {
            m.log("else 2")
        }
        if e.Params["b"].(bool) {
            compartment := NewBranchCompartment(BranchState_F1)
            m._transition_(compartment)
            return
        } else {
            compartment := NewBranchCompartment(BranchState_F2)
            m._transition_(compartment)
            return
        }
        
        return
    case "OnInt":
        if e.Params["i"].(int16) > 5 {
            m.log("> 5")
        } else {
            m.log("<= 5")
        }
        if e.Params["i"].(int16) < 10 {
            m.log("< 10")
        } else {
            m.log(">= 10")
        }
        if e.Params["i"].(int16) == 7 {
            m.log("== 7")
            compartment := NewBranchCompartment(BranchState_F1)
            m._transition_(compartment)
            return
        } else {
            m.log("!= 7")
            compartment := NewBranchCompartment(BranchState_F2)
            m._transition_(compartment)
            return
        }
        
        return
    }
}

func (m *branchStruct) _BranchState_NegatedIf_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "OnBool":
        if !(e.Params["b"].(bool)) {
            m.log("then 1")
        } else {
        }
        if !(e.Params["b"].(bool)) {
        } else {
            m.log("else 1")
        }
        if !(e.Params["b"].(bool)) {
            m.log("then 2")
        } else {
            m.log("else 2")
        }
        if !(e.Params["b"].(bool)) {
            compartment := NewBranchCompartment(BranchState_F1)
            m._transition_(compartment)
            return
        } else {
            compartment := NewBranchCompartment(BranchState_F2)
            m._transition_(compartment)
            return
        }
        
        return
    case "OnInt":
        if !(e.Params["i"].(int16) >= 5) {
            m.log("< 5")
        } else {
            m.log(">= 5")
        }
        if !(e.Params["i"].(int16) <= 10) {
            m.log("> 10")
        } else {
            m.log("<= 10")
        }
        if !(e.Params["i"].(int16) != 7) {
            m.log("== 7")
            compartment := NewBranchCompartment(BranchState_F1)
            m._transition_(compartment)
            return
        } else {
            m.log("!= 7")
            compartment := NewBranchCompartment(BranchState_F2)
            m._transition_(compartment)
            return
        }
        
        return
    }
}

func (m *branchStruct) _BranchState_Precedence_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "OnInt":
        if -e.Params["i"].(int16) >= 0 && -e.Params["i"].(int16) <= 5 {
            m.log("then 1")
        } else {
            m.log("else 1")
        }
        if !(e.Params["i"].(int16) >= -5 && e.Params["i"].(int16) <= 5) && (e.Params["i"].(int16) >= -10 && e.Params["i"].(int16) <= 10) {
            m.log("then 2")
        } else {
            m.log("else 2")
        }
        if e.Params["i"].(int16) >= 0 && e.Params["i"].(int16) <= 5 || e.Params["i"].(int16) >= 10 && e.Params["i"].(int16) <= 20 {
            m.log("then 3")
        } else {
            m.log("else 3")
        }
        if !((e.Params["i"].(int16) < 0 || e.Params["i"].(int16) > 10) && e.Params["i"].(int16) + 5 < 20) {
            m.log("then 4")
        } else {
            m.log("else 4")
        }
        
        return
    }
}

func (m *branchStruct) _BranchState_NestedIf_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "OnInt":
        if e.Params["i"].(int16) > 0 {
            m.log("> 0")
            if e.Params["i"].(int16) < 100 {
                m.log("< 100")
                compartment := NewBranchCompartment(BranchState_F1)
                m._transition_(compartment)
                return
            } else {
                m.log(">= 100")
            }
        } else {
            m.log("<= 0")
            if e.Params["i"].(int16) > -10 {
                m.log("> -10")
            } else {
                m.log("<= -10")
                compartment := NewBranchCompartment(BranchState_F2)
                m._transition_(compartment)
                return
            }
        }
        
        return
    }
}

func (m *branchStruct) _BranchState_GuardedTransition_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "OnInt":
        if e.Params["i"].(int16) > 100 {
            m.log("-> $F1")
            compartment := NewBranchCompartment(BranchState_F1)
            m._transition_(compartment)
            return
        } else {
        }
        if !(e.Params["i"].(int16) > 10) {
        } else {
            m.log("-> $F2")
            compartment := NewBranchCompartment(BranchState_F2)
            m._transition_(compartment)
            return
        }
        m.log("-> $F3")
        compartment := NewBranchCompartment(BranchState_F3)
        m._transition_(compartment)
        
        return
    }
}

func (m *branchStruct) _BranchState_NestedGuardedTransition_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "OnInt":
        if e.Params["i"].(int16) > 10 {
            if e.Params["i"].(int16) > 100 {
                m.log("-> $F1")
                compartment := NewBranchCompartment(BranchState_F1)
                m._transition_(compartment)
                return
            } else {
            }
            if e.Params["i"].(int16) > 50 {
            } else {
                m.log("-> $F2")
                compartment := NewBranchCompartment(BranchState_F2)
                m._transition_(compartment)
                return
            }
        } else {
        }
        m.log("-> $F3")
        compartment := NewBranchCompartment(BranchState_F3)
        m._transition_(compartment)
        
        return
    }
}

func (m *branchStruct) _BranchState_F1_(e *framelang.FrameEvent) {
    switch e.Msg {
    }
}

func (m *branchStruct) _BranchState_F2_(e *framelang.FrameEvent) {
    switch e.Msg {
    }
}

func (m *branchStruct) _BranchState_F3_(e *framelang.FrameEvent) {
    switch e.Msg {
    }
}

//=============== Machinery and Mechanisms ==============//

func (m *branchStruct) _transition_(compartment *BranchCompartment) {
    m._nextCompartment_ = compartment
}

func (m *branchStruct) _do_transition_(nextCompartment *BranchCompartment) {
    m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
    m._compartment_ = nextCompartment
    m._mux_(&framelang.FrameEvent{Msg: ">", Params: m._compartment_.EnterArgs, Ret: nil})
}

//===================== Actions Block ===================//


/********************************************************

// Unimplemented Actions

func (m *branchStruct) log(msg string)  {}

********************************************************/

//=============== Compartment ==============//

type BranchCompartment struct {
    State BranchState
    StateArgs map[string]interface{}
    StateVars map[string]interface{}
    EnterArgs map[string]interface{}
    ExitArgs map[string]interface{}
    _forwardEvent_ *framelang.FrameEvent
}

func NewBranchCompartment(state BranchState) *BranchCompartment {
    c := &BranchCompartment{State: state}
    c.StateArgs = make(map[string]interface{})
    c.StateVars = make(map[string]interface{})
    c.EnterArgs = make(map[string]interface{})
    c.ExitArgs = make(map[string]interface{})
    return c
}

File: framec_tests/python/src/branch/branch.py
#Emitted from framec_v0.11.2



from framelang.framelang import FrameEvent



class FrameEvent:
    def __init__(self, message, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class Branch:
    
    
    # ==================== System Factory =================== #
    
    def __init__(self):
        
         # Create and initialize start state compartment.
        
        next_compartment = None
        next_compartment = BranchCompartment('__branch_state_I', next_compartment)
        self.__compartment = next_compartment
        self.__next_compartment = None
        self.return_stack = [None]
        
        # Initialize domain
        
        self.tape  = []
        
        # Send system start event
        frame_event = FrameEvent(">", None)
        self.__kernel(frame_event)
    
    # ==================== Interface Block ================== #
    
    def A(self,):
        self.return_stack.append(None)
        __e = FrameEvent("A",None)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    def B(self,):
        self.return_stack.append(None)
        __e = FrameEvent("B",None)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    def C(self,):
        self.return_stack.append(None)
        __e = FrameEvent("C",None)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    def D(self,):
        self.return_stack.append(None)
        __e = FrameEvent("D",None)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    def E(self,):
        self.return_stack.append(None)
        __e = FrameEvent("E",None)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    def F(self,):
        self.return_stack.append(None)
        __e = FrameEvent("F",None)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    def OnBool(self,b: bool):
        parameters = {}
        parameters["b"] = b
        self.return_stack.append(None)
        __e = FrameEvent("OnBool",parameters)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    def OnInt(self,i: int):
        parameters = {}
        parameters["i"] = i
        self.return_stack.append(None)
        __e = FrameEvent("OnInt",parameters)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    # ===================== Machine Block =================== #
    
    # ----------------------------------------
    # $I
    
    def __branch_state_I(self, __e, compartment):
        if __e._message == "A":
            next_compartment = None
            next_compartment = BranchCompartment('__branch_state_SimpleIf', next_compartment)
            self.__transition(next_compartment)
            return
        elif __e._message == "B":
            next_compartment = None
            next_compartment = BranchCompartment('__branch_state_NegatedIf', next_compartment)
            self.__transition(next_compartment)
            return
        elif __e._message == "C":
            next_compartment = None
            next_compartment = BranchCompartment('__branch_state_Precedence', next_compartment)
            self.__transition(next_compartment)
            return
        elif __e._message == "D":
            next_compartment = None
            next_compartment = BranchCompartment('__branch_state_NestedIf', next_compartment)
            self.__transition(next_compartment)
            return
        elif __e._message == "E":
            next_compartment = None
            next_compartment = BranchCompartment('__branch_state_GuardedTransition', next_compartment)
            self.__transition(next_compartment)
            return
        elif __e._message == "F":
            next_compartment = None
            next_compartment = BranchCompartment('__branch_state_NestedGuardedTransition', next_compartment)
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $SimpleIf
    
    def __branch_state_SimpleIf(self, __e, compartment):
        if __e._message == "OnBool":
            if  __e._parameters["b"]:
                self.log_do("then 1")
            else:
                pass
            
            if  __e._parameters["b"]:
                pass
            else:
                self.log_do("else 1")
            
            if  __e._parameters["b"]:
                self.log_do("then 2")
            else:
                self.log_do("else 2")
            
            if  __e._parameters["b"]:
                next_compartment = None
                next_compartment = BranchCompartment('__branch_state_F1', next_compartment)
                self.__transition(next_compartment)
                return
            else:
                next_compartment = None
                next_compartment = BranchCompartment('__branch_state_F2', next_compartment)
                self.__transition(next_compartment)
                return
            
            return
        elif __e._message == "OnInt":
            if  __e._parameters["i"] > 5:
                self.log_do("> 5")
            else:
                self.log_do("<= 5")
            
            if  __e._parameters["i"] < 10:
                self.log_do("< 10")
            else:
                self.log_do(">= 10")
            
            if  __e._parameters["i"] == 7:
                self.log_do("== 7")
                next_compartment = None
                next_compartment = BranchCompartment('__branch_state_F1', next_compartment)
                self.__transition(next_compartment)
                return
            else:
                self.log_do("!= 7")
                next_compartment = None
                next_compartment = BranchCompartment('__branch_state_F2', next_compartment)
                self.__transition(next_compartment)
                return
            
            return
    
    # ----------------------------------------
    # $NegatedIf
    
    def __branch_state_NegatedIf(self, __e, compartment):
        if __e._message == "OnBool":
            if  not (__e._parameters["b"]):
                self.log_do("then 1")
            else:
                pass
            
            if  not (__e._parameters["b"]):
                pass
            else:
                self.log_do("else 1")
            
            if  not (__e._parameters["b"]):
                self.log_do("then 2")
            else:
                self.log_do("else 2")
            
            if  not (__e._parameters["b"]):
                next_compartment = None
                next_compartment = BranchCompartment('__branch_state_F1', next_compartment)
                self.__transition(next_compartment)
                return
            else:
                next_compartment = None
                next_compartment = BranchCompartment('__branch_state_F2', next_compartment)
                self.__transition(next_compartment)
                return
            
            return
        elif __e._message == "OnInt":
            if  not (__e._parameters["i"] >= 5):
                self.log_do("< 5")
            else:
                self.log_do(">= 5")
            
            if  not (__e._parameters["i"] <= 10):
                self.log_do("> 10")
            else:
                self.log_do("<= 10")
            
            if  not (__e._parameters["i"] != 7):
                self.log_do("== 7")
                next_compartment = None
                next_compartment = BranchCompartment('__branch_state_F1', next_compartment)
                self.__transition(next_compartment)
                return
            else:
                self.log_do("!= 7")
                next_compartment = None
                next_compartment = BranchCompartment('__branch_state_F2', next_compartment)
                self.__transition(next_compartment)
                return
            
            return
    
    # ----------------------------------------
    # $Precedence
    
    def __branch_state_Precedence(self, __e, compartment):
        if __e._message == "OnInt":
            if  -__e._parameters["i"] >= 0 and -__e._parameters["i"] <= 5:
                self.log_do("then 1")
            else:
                self.log_do("else 1")
            
            if   not (__e._parameters["i"] >= -5 and __e._parameters["i"] <= 5) and (__e._parameters["i"] >= -10 and __e._parameters["i"] <= 10):
                self.log_do("then 2")
            else:
                self.log_do("else 2")
            
            if  __e._parameters["i"] >= 0 and __e._parameters["i"] <= 5 or __e._parameters["i"] >= 10 and __e._parameters["i"] <= 20:
                self.log_do("then 3")
            else:
                self.log_do("else 3")
            
            if  not ((__e._parameters["i"] < 0 or __e._parameters["i"] > 10) and __e._parameters["i"] + 5 < 20):
                self.log_do("then 4")
            else:
                self.log_do("else 4")
            
            return
    
    # ----------------------------------------
    # $NestedIf
    
    def __branch_state_NestedIf(self, __e, compartment):
        if __e._message == "OnInt":
            if  __e._parameters["i"] > 0:
                self.log_do("> 0")
                if  __e._parameters["i"] < 100:
                    self.log_do("< 100")
                    next_compartment = None
                    next_compartment = BranchCompartment('__branch_state_F1', next_compartment)
                    self.__transition(next_compartment)
                    return
                else:
                    self.log_do(">= 100")
                
            else:
                self.log_do("<= 0")
                if  __e._parameters["i"] > -10:
                    self.log_do("> -10")
                else:
                    self.log_do("<= -10")
                    next_compartment = None
                    next_compartment = BranchCompartment('__branch_state_F2', next_compartment)
                    self.__transition(next_compartment)
                    return
                
            
            return
    
    # ----------------------------------------
    # $GuardedTransition
    
    def __branch_state_GuardedTransition(self, __e, compartment):
        if __e._message == "OnInt":
            if  __e._parameters["i"] > 100:
                self.log_do("-> $F1")
                next_compartment = None
                next_compartment = BranchCompartment('__branch_state_F1', next_compartment)
                self.__transition(next_compartment)
                return
            else:
                pass
            
            if  not (__e._parameters["i"] > 10):
                pass
            else:
                self.log_do("-> $F2")
                next_compartment = None
                next_compartment = BranchCompartment('__branch_state_F2', next_compartment)
                self.__transition(next_compartment)
                return
            
            self.log_do("-> $F3")
            next_compartment = None
            next_compartment = BranchCompartment('__branch_state_F3', next_compartment)
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $NestedGuardedTransition
    
    def __branch_state_NestedGuardedTransition(self, __e, compartment):
        if __e._message == "OnInt":
            if  __e._parameters["i"] > 10:
                if  __e._parameters["i"] > 100:
                    self.log_do("-> $F1")
                    next_compartment = None
                    next_compartment = BranchCompartment('__branch_state_F1', next_compartment)
                    self.__transition(next_compartment)
                    return
                else:
                    pass
                
                if  __e._parameters["i"] > 50:
                    pass
                else:
                    self.log_do("-> $F2")
                    next_compartment = None
                    next_compartment = BranchCompartment('__branch_state_F2', next_compartment)
                    self.__transition(next_compartment)
                    return
                
            else:
                pass
            
            self.log_do("-> $F3")
            next_compartment = None
            next_compartment = BranchCompartment('__branch_state_F3', next_compartment)
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $F1
    
    def __branch_state_F1(self, __e, compartment):
        pass
        
    
    # ----------------------------------------
    # $F2
    
    def __branch_state_F2(self, __e, compartment):
        pass
        
    
    # ----------------------------------------
    # $F3
    
    def __branch_state_F3(self, __e, compartment):
        pass
        
    
    # ===================== Actions Block =================== #
    
    def log_do(self,msg: str):
        raise NotImplementedError
    
    # ==================== System Runtime =================== #
    
    def __kernel(self, __e):
        
        # send event to current state
        self.__router(__e)
        
        # loop until no transitions occur
        while self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            
            # exit current state
            self.__router(FrameEvent( "<", self.__compartment.exit_args))
            # change state
            self.__compartment = next_compartment
            
            if next_compartment.forward_event is None:
                # send normal enter event
                self.__router(FrameEvent(">", self.__compartment.enter_args))
            else: # there is a forwarded event
                if next_compartment.forward_event._message == ">":
                    # forwarded event is enter event
                    self.__router(next_compartment.forward_event)
                else:
                    # forwarded event is not enter event
                    # send normal enter event
                    self.__router(FrameEvent(">", self.__compartment.enter_args))
                    # and now forward event to new, intialized state
                    self.__router(next_compartment.forward_event)
                next_compartment.forward_event = None
                
    
    def __router(self, __e):
        if self.__compartment.state == '__branch_state_I':
            self.__branch_state_I(__e, self.__compartment)
        elif self.__compartment.state == '__branch_state_SimpleIf':
            self.__branch_state_SimpleIf(__e, self.__compartment)
        elif self.__compartment.state == '__branch_state_NegatedIf':
            self.__branch_state_NegatedIf(__e, self.__compartment)
        elif self.__compartment.state == '__branch_state_Precedence':
            self.__branch_state_Precedence(__e, self.__compartment)
        elif self.__compartment.state == '__branch_state_NestedIf':
            self.__branch_state_NestedIf(__e, self.__compartment)
        elif self.__compartment.state == '__branch_state_GuardedTransition':
            self.__branch_state_GuardedTransition(__e, self.__compartment)
        elif self.__compartment.state == '__branch_state_NestedGuardedTransition':
            self.__branch_state_NestedGuardedTransition(__e, self.__compartment)
        elif self.__compartment.state == '__branch_state_F1':
            self.__branch_state_F1(__e, self.__compartment)
        elif self.__compartment.state == '__branch_state_F2':
            self.__branch_state_F2(__e, self.__compartment)
        elif self.__compartment.state == '__branch_state_F3':
            self.__branch_state_F3(__e, self.__compartment)
        
    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment
    
    def state_info(self):
        return self.__compartment.state
        

# ===================== Compartment =================== #

class BranchCompartment:

    def __init__(self,state,parent_compartment):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment
    